drivers/stepper/gpio_stepper_controller.c

```diff
@@ -0,0 +1,337 @@
+#include <zephyr/kernel.h>
+#include <zephyr/sys_clock.h>
+#include <zephyr/drivers/stepper.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/__assert.h>
+
+LOG_MODULE_REGISTER(gpio_stepper_motor_controller, CONFIG_STEPPER_LOG_LEVEL);
+
+#define MAX_MICRO_STEP_RES STEPPER_MICRO_STEP_2
+#define NUM_CONTROL_PINS   4
+
+static const uint8_t gpio_stepper_half_step_lookup_table[NUM_CONTROL_PINS *
+							 MAX_MICRO_STEP_RES][NUM_CONTROL_PINS] = {
+	{1, 1, 0, 0}, {0, 1, 0, 0}, {0, 1, 1, 0}, {0, 0, 1, 0},
+	{0, 0, 1, 1}, {0, 0, 0, 1}, {1, 0, 0, 1}, {1, 0, 0, 0}};
+
+struct gpio_stepper_motor_controller_config {
+	const struct gpio_dt_spec *control_pins;
+};
+
+struct gpio_stepper_motor_controller_data {
+	const struct device *dev;
+	struct k_spinlock lock;
+	enum stepper_direction direction;
+	enum stepper_run_mode run_mode;
+	uint8_t step_gap;
+	uint8_t coil_charge;
+	struct k_work_delayable stepper_dwork;
+	struct k_poll_signal *async_signal;
+	int32_t actual_position;
+	uint32_t delay_in_us;
+	int32_t step_count;
+};
+
+static int stepper_motor_set_coil_charge(const struct device *dev)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+	const struct gpio_stepper_motor_controller_config *config = dev->config;
+
+	for (int i = 0; i < NUM_CONTROL_PINS; i++) {
+		gpio_pin_set_dt(&config->control_pins[i],
+				gpio_stepper_half_step_lookup_table[data->coil_charge][i]);
+	}
+	return 0;
+}
+
+static void update_coil_charge(const struct device *dev)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	if (data->direction == STEPPER_DIRECTION_POSITIVE) {
+		data->coil_charge = data->coil_charge == 0
+					    ? NUM_CONTROL_PINS * MAX_MICRO_STEP_RES - data->step_gap
+					    : data->coil_charge - data->step_gap;
+		data->actual_position++;
+	} else if (data->direction == STEPPER_DIRECTION_NEGATIVE) {
+		data->coil_charge =
+			data->coil_charge == NUM_CONTROL_PINS * MAX_MICRO_STEP_RES - data->step_gap
+				? 0
+				: data->coil_charge + data->step_gap;
+		data->actual_position--;
+	}
+}
+
+static void update_remaining_steps(struct gpio_stepper_motor_controller_data *data)
+{
+	if (data->step_count > 0) {
+		data->step_count--;
+	} else if (data->step_count < 0) {
+		data->step_count++;
+	} else {
+		if (data->async_signal) {
+			k_poll_signal_raise(data->async_signal, STEPPER_SIGNAL_STEPS_COMPLETED);
+		}
+	}
+}
+
+static void update_direction_from_step_count(struct gpio_stepper_motor_controller_data *data)
+{
+	if (data->step_count > 0) {
+		data->direction = STEPPER_DIRECTION_POSITIVE;
+	} else if (data->step_count < 0) {
+		data->direction = STEPPER_DIRECTION_NEGATIVE;
+	} else {
+		LOG_ERR("Step count is zero");
+	}
+}
+
+static void position_mode_task(const struct device *dev)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	stepper_motor_set_coil_charge(dev);
+	update_coil_charge(dev);
+	update_remaining_steps(dev->data);
+	k_work_reschedule(&data->stepper_dwork, K_USEC(data->delay_in_us));
```

Comment:
`delay_in_us`  initializes as 0, so if one does not set the velocity this results in the task being schedule instantly, tried the code and it seems like it was spinning as well, starving the main task out of resources, should the move API have a check to make sure that the delay has been set?